README.md
# onionpipe

Onion addresses for anything.

`onionpipe` forwards ports on the local host to remote Onion addresses as Tor
hidden services and vice-versa.

## Why would I want to use this?

onionpipe is a decentralized way to create virtually unstoppable global network
tunnels.

For example, you might want to securely publish and access a personal service
from anywhere in the world, across all sorts of network obstructions -- your
ISP doesn't allow ingress traffic to your home lab, your clients might be in
heavily firewalled environments (public WiFi, mobile tether), etc.

With onionpipe, that service doesn't need a public IPv4 or IPv6 ingress. You
can publish services with a globally-unique persistent onion address, and share
access securely and privately to your own allowlist of authorized keys.

## How do I install it?

On Linux (and probably macOS? untested atm):

    cargo install onionpipe

## What can I do with it right now?

onionpipe sets up socket forwarding tunnels. It's like `socat(1)`, for onions.

### Export services on local networks to onion addresses

Export localhost port 8000 to a temporary, one-time remote onion address. Default is port 80 on the onion service.

```
onionpipe 8000
```

The forwarding address is printed to the log output:

```
Feb 26 09:41:18.267 [notice] Tor 0.4.7.10 running on Linux with Libevent 2.1.12-stable, OpenSSL 1.1.1t, Zlib 1.2.11, Liblzma N/A, Libzstd N/A and Glibc 2.35 as libc.
Feb 26 09:41:18.267 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://support.torproject.org/faq/staying-anonymous/
Feb 26 09:41:18.267 [notice] Configuration file "/home/c/.torrc" not present, using reasonable defaults.
Feb 26 09:41:18.271 [notice] Opening Socks listener on /run/user/1000/.tmpfviq7N/data/socks.sock
Feb 26 09:41:18.271 [notice] Opened Socks listener connection (ready) on /run/user/1000/.tmpfviq7N/data/socks.sock
Feb 26 09:41:18.271 [notice] Opening Control listener on /run/user/1000/.tmpfviq7N/data/control.sock
Feb 26 09:41:18.271 [notice] Opened Control listener connection (ready) on /run/user/1000/.tmpfviq7N/data/control.sock
forward 127.0.0.1:8000 => pqksfxbpraiwklpx7ihu7yu7vlpkpromqojyn6goo2fl6wemi4dkieqd.onion:80
```

Port forwarding can be mapped. This exports localhost port 8443 to temporary remote onion port 443. `~` is shorthand
for the forward between source~destination.

```
onionpipe 8443~443
```

Local addresses may be bound. This forwards a specific interface address to an onion:

```
onionpipe 10.0.0.7:8443~443
```

### Persistent onion addresses

```
onionpipe 8000@my-app
```

### Import onion services


This imports an Onion site to a local listener on port 8000.

```
onionpipe ddosxlvzzow7scc7egy75gpke54hgbg2frahxzaw6qq5osnzm7wistid.onion~8000
```

Import an Onion site to a specific address. Useful for setting up an intranet or clearnet ingress to the onion service.

```
onionpipe ddosxlvzzow7scc7egy75gpke54hgbg2frahxzaw6qq5osnzm7wistid.onion~0.0.0.0:8000
```

### Config file operation

All the above and more can be expressed with a JSON configuration file. See [Config](https://docs.rs/onionpipe/0.3.0/onionpipe/config/struct.Config.html) Rust docs and [an example config.json](examples/config.json) for details.

```
onionpipe --config config.json
```

## TODOs

- Security review. Rust code review, I'm kind of new to the language.
- CLI compatibility with the [Go implementation](https://github.com/cmars/onionpipe). What's still missing?
  - Client authentication & key management
  - More Tor options like anonymous vs fast, bridge support. Vanguard integration.
  - UNIX socket support. Doable but a dependency will need some enhancement (torut)
- Cross-platform distribution of the above: Linux, macOS, Windows on popular architectures
  - Distributions on Docker, NixOS (flake), Homebrew, maybe Choco?

## More ideas!

- GUI front-end, possibly based on Tauri
- cwtch integration
- daemon mode & forwarding control API
- Kubernetes CRD: OnionService (which could use the control API)
- Arti-based fork, when Arti supports hidden services



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.4.0 (2023-05-06)

### Chore

 - <csr-id-66945d12d82eae2161ba279dd804e06164d57389/> unlimited checkout depth
   Trying to fix smart-release action
 - <csr-id-51e1e408e6601e0a4537d870e0130dd9334f69e9/> checkout with history for smart-release

### Bug Fixes

 - <csr-id-a8e46cd0bef575fd582e9e9f3ad19bc9cb6d66ce/> add release action

### New Features (BREAKING)

 - <csr-id-d62e283fbe28204f72e1cb58666de0e3b82c5f0c/> add service management cli commands
   Add CLI subcommands for adding (idempotent), listing and deleting
   service v3 keys.

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 4 commits contributed to the release.
 - 5 days passed between releases.
 - 4 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 2 unique issues were worked on: [#11](https://github.com/cmars/onionpipe-rs/issues/11), [#12](https://github.com/cmars/onionpipe-rs/issues/12)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#11](https://github.com/cmars/onionpipe-rs/issues/11)**
    - Add service management cli commands ([`d62e283`](https://github.com/cmars/onionpipe-rs/commit/d62e283fbe28204f72e1cb58666de0e3b82c5f0c))
 * **[#12](https://github.com/cmars/onionpipe-rs/issues/12)**
    - Add release action ([`a8e46cd`](https://github.com/cmars/onionpipe-rs/commit/a8e46cd0bef575fd582e9e9f3ad19bc9cb6d66ce))
 * **Uncategorized**
    - Unlimited checkout depth ([`66945d1`](https://github.com/cmars/onionpipe-rs/commit/66945d12d82eae2161ba279dd804e06164d57389))
    - Checkout with history for smart-release ([`51e1e40`](https://github.com/cmars/onionpipe-rs/commit/51e1e408e6601e0a4537d870e0130dd9334f69e9))
</details>

## v0.3.2 (2023-04-30)

<csr-id-401c60cf746b727c35e60bdeb29e280f0449ff8e/>

### Chore

 - <csr-id-401c60cf746b727c35e60bdeb29e280f0449ff8e/> derive config from cli expressions
   Drive onionpipe builder entirely from config, deriving a dynamic config
   from the CLI if a file is not specified.

### New Features

 - <csr-id-70a15cabbbd8fdbdab8e61da9c36fffaf0d7cead/> secret key store for service and client auth keys

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release over the course of 42 calendar days.
 - 63 days passed between releases.
 - 2 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 2 unique issues were worked on: [#8](https://github.com/cmars/onionpipe-rs/issues/8), [#9](https://github.com/cmars/onionpipe-rs/issues/9)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#8](https://github.com/cmars/onionpipe-rs/issues/8)**
    - Derive config from cli expressions ([`401c60c`](https://github.com/cmars/onionpipe-rs/commit/401c60cf746b727c35e60bdeb29e280f0449ff8e))
 * **[#9](https://github.com/cmars/onionpipe-rs/issues/9)**
    - Secret key store for service and client auth keys ([`70a15ca`](https://github.com/cmars/onionpipe-rs/commit/70a15cabbbd8fdbdab8e61da9c36fffaf0d7cead))
 * **Uncategorized**
    - Release onionpipe v0.3.2 ([`35b428e`](https://github.com/cmars/onionpipe-rs/commit/35b428ec0ded06c18d3439aac7573753730a58e5))
</details>

## v0.3.1 (2023-02-26)

<csr-id-fe374c36842ce95366842e3b141a25147b94fd07/>

### Chore

 - <csr-id-fe374c36842ce95366842e3b141a25147b94fd07/> README updates
   Updating README to reflect state of latest 0.3.0 release

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#6](https://github.com/cmars/onionpipe-rs/issues/6)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#6](https://github.com/cmars/onionpipe-rs/issues/6)**
    - README updates ([`fe374c3`](https://github.com/cmars/onionpipe-rs/commit/fe374c36842ce95366842e3b141a25147b94fd07))
 * **Uncategorized**
    - Release onionpipe v0.3.1 ([`446a55c`](https://github.com/cmars/onionpipe-rs/commit/446a55c57f9e26322fa2d35abd3cad5c7a998e0d))
</details>

## v0.3.0 (2023-02-25)

### New Features

 - <csr-id-b5554e040254cd4639b741f3d5b256857c126662/> parse forward expressions from CLI
   Parse multiple onionpipe-style expressions from command line args.

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 5 days passed between releases.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#5](https://github.com/cmars/onionpipe-rs/issues/5)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#5](https://github.com/cmars/onionpipe-rs/issues/5)**
    - Parse forward expressions from CLI ([`b5554e0`](https://github.com/cmars/onionpipe-rs/commit/b5554e040254cd4639b741f3d5b256857c126662))
 * **Uncategorized**
    - Release onionpipe v0.3.0 ([`bf2c441`](https://github.com/cmars/onionpipe-rs/commit/bf2c441c5a3cc0bfc813ee86abe673f57cc097a2))
</details>

## v0.2.2 (2023-02-20)

<csr-id-959c781bcca5a17d23edc388734f3ec232dfc3aa/>

### Chore

 - <csr-id-959c781bcca5a17d23edc388734f3ec232dfc3aa/> add gh to nix shell
   gh needed for Github releases

### New Features

 - <csr-id-28c832efe646a3c51d912ce9d2de487491b8065d/> forward expression parser
   Parse onionpipe forwarding expressions.

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release.
 - 7 days passed between releases.
 - 2 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#4](https://github.com/cmars/onionpipe-rs/issues/4)

### Commit Details


src/parse.rs
// Some parsing functions adapted from https://github.com/zupzup/rust-nom-parsing, which is
// licensed:
//
// Copyright 2020 Mario Zupan <mario@zupzup.org>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub mod parse {

    use std::str::FromStr;

    use nom::{
        branch::alt,
        bytes::complete::tag,
        character::complete::{alphanumeric1, one_of},
        combinator::{eof, opt},
        error::{context, VerboseError},
        multi::{count, many_m_n, separated_list1},
        sequence::{preceded, terminated, tuple},
        Err as NomErr, Finish, IResult,
    };

    #[derive(Debug, PartialEq, Eq)]
    pub struct ParseError(String);

    #[derive(Debug, PartialEq, Eq)]
    pub struct ExportLocalTCPAddr {
        host: Option<Host>,
        port: u16,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Host {
        HOST(String),
        IP4([u8; 4]),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum ExportLocalAddr {
        TCP(ExportLocalTCPAddr),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct ExportRemoteAddr {
        onion_alias: Option<String>,
        ports: Vec<u16>,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct ExportForward {
        local: ExportLocalAddr,
        remote: Option<ExportRemoteAddr>,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct ImportRemoteAddr {
        onion: String,
        port: Option<u16>,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum ImportLocalAddr {
        TCP(ImportLocalTCPAddr),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct ImportLocalTCPAddr {
        host: Option<Host>,
        port: Option<u16>,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct ImportForward {
        remote: ImportRemoteAddr,
        local: Option<ImportLocalAddr>,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Forward {
        Export(ExportForward),
        Import(ImportForward),
    }

    impl FromStr for Forward {
        type Err = ParseError;

        fn from_str(s: &str) -> Result<Self, Self::Err> {
            match forward(s).finish() {
                Ok((_, forward)) => Ok(forward),
                Err(e) => Err(ParseError(e.to_string())),
            }
        }
    }

    type Res<T, U> = IResult<T, U, VerboseError<T>>;

    fn forward(input: &str) -> Res<&str, Forward> {
        context(
            "forward",
            tuple((alt((import_forward, export_forward)), eof)),
        )(input)
        .map(|(next_input, res)| (next_input, res.0))
    }

    fn export_forward(input: &str) -> Res<&str, Forward> {
        context(
            "export forward",
            tuple((local_tcp_addr, opt(preceded(tag("~"), export_remote_addr)))),
        )(input)
        .map(|(next_input, res)| {
            let result: Forward = Forward::Export(ExportForward {
                local: ExportLocalAddr::TCP(res.0),
                remote: res.1,
            });
            (next_input, result)
        })
    }

    fn export_remote_addr(input: &str) -> Res<&str, ExportRemoteAddr> {
        context(
            "export remote addr",
            tuple((opt(terminated(alphanumeric1, tag(":"))), ports)),
        )(input)
        .map(|(next_input, res)| {
            let result: ExportRemoteAddr = ExportRemoteAddr {
                onion_alias: res.0.map(|s| s.to_string()),
                ports: res.1,
            };
            (next_input, result)
        })
    }

    fn local_tcp_addr(input: &str) -> Res<&str, ExportLocalTCPAddr> {
        context(
            "local tcp addr",
            tuple((opt(terminated(ip, tag(":"))), port)),
        )(input)
        .map(|(next_input, res)| {
            let result: ExportLocalTCPAddr = ExportLocalTCPAddr {
                host: res.0,
                port: res.1,
            };
            (next_input, result)
        })
    }

    fn import_forward(input: &str) -> Res<&str, Forward> {
        context(
            "import forward",
            tuple((
                import_remote_addr,
                opt(preceded(tag("~"), import_local_addr)),
            )),
        )(input)
        .map(|(next_input, res)| {
            let result: Forward = Forward::Import(ImportForward {
                remote: res.0,
                local: res.1,
            });
            (next_input, result)
        })
    }

    fn import_remote_addr(input: &str) -> Res<&str, ImportRemoteAddr> {
        context(
            "import remote addr",
            tuple((onion, opt(preceded(tag(":"), port)))),
        )(input)
        .map(|(next_input, res)| {
            let result: ImportRemoteAddr = ImportRemoteAddr {
                onion: res.0.to_owned(),
                port: res.1,
            };
            (next_input, result)
        })
    }

    fn onion(input: &str) -> Res<&str, String> {
        context("onion", terminated(alphanumeric1, tag(".onion")))(input)
            .map(|(next_input, res)| (next_input, res.to_owned()))
    }

    fn import_local_addr(input: &str) -> Res<&str, ImportLocalAddr> {
        context(
            "import local addr",
            alt((
                import_local_addr_host_port,
                import_local_addr_host_only,
                import_local_addr_port_only,
            )),
        )(input)
        .map(|(next_input, res)| {
            let result: ImportLocalAddr = ImportLocalAddr::TCP(res);
            (next_input, result)
        })
    }

    fn import_local_addr_host_port(input: &str) -> Res<&str, ImportLocalTCPAddr> {
        context("import local addr host:port", tuple((ip, tag(":"), port)))(input).map(
            |(next_input, res)| {
                (
                    next_input,
                    ImportLocalTCPAddr {
                        host: Some(res.0),
                        port: Some(res.2),
                    },
                )
            },
        )
    }

    fn import_local_addr_host_only(input: &str) -> Res<&str, ImportLocalTCPAddr> {
        context("import local addr host only", ip)(input).map(|(next_input, res)| {
            (
                next_input,
                ImportLocalTCPAddr {
                    host: Some(res),
                    port: None,
                },
            )
        })
    }

    fn import_local_addr_port_only(input: &str) -> Res<&str, ImportLocalTCPAddr> {
        context("import local addr host only", port)(input).map(|(next_input, res)| {
            (
                next_input,
                ImportLocalTCPAddr {
                    host: None,
                    port: Some(res),
                },
            )
        })
    }

    fn ip(input: &str) -> Res<&str, Host> {
        context(
            "ip",
            tuple((count(terminated(ip_num, tag(".")), 3), ip_num)),
        )(input)
        .map(|(next_input, res)| {
            let mut result: [u8; 4] = [0, 0, 0, 0];
            res.0
                .into_iter()
                .enumerate()
                .for_each(|(i, v)| result[i] = v);
            result[3] = res.1;
            (next_input, Host::IP4(result))
        })
    }

    fn ip_num(input: &str) -> Res<&str, u8> {
        context("ip number", n_to_m_digits(1, 3))(input).and_then(|(next_input, result)| {
            match result.parse::<u8>() {
                Ok(n) => Ok((next_input, n)),
                Err(_) => Err(NomErr::Error(VerboseError { errors: vec![] })),
            }
        })
    }

    fn port(input: &str) -> Res<&str, u16> {
        context("port", n_to_m_digits(2, 5))(input).and_then(|(next_input, res)| {
            match res.parse::<u16>() {
                Ok(n) => Ok((next_input, n)),
                Err(_) => Err(NomErr::Error(VerboseError { errors: vec![] })),
            }
        })
    }

    fn ports(input: &str) -> Res<&str, Vec<u16>> {
        context("ports", separated_list1(tag(","), port))(input).map(|(next_input, res)| {
            let mut result: Vec<u16> = vec![0; res.len()];
            res.into_iter().enumerate().for_each(|(i, v)| result[i] = v);
            (next_input, result)
        })
    }

    fn n_to_m_digits<'a>(n: usize, m: usize) -> impl FnMut(&'a str) -> Res<&str, String> {
        move |input| {
            many_m_n(n, m, one_of("0123456789"))(input)
                .map(|(next_input, result)| (next_input, result.into_iter().collect()))
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;
        use nom::{
            error::{ErrorKind, VerboseError, VerboseErrorKind},
            Err as NomErr,
        };

        #[test]
        fn test_local_tcp_addr() {
            assert_eq!(
                local_tcp_addr("127.0.0.1:4567~"),
                Ok((
                    "~",
                    ExportLocalTCPAddr {
                        host: Some(Host::IP4([127, 0, 0, 1])),
                        port: 4567
                    }
                ))
            );
            assert_eq!(
                local_tcp_addr("6667~"),
                Ok((
                    "~",
                    ExportLocalTCPAddr {
                        host: None,
                        port: 6667
                    }
                ))
            );
        }

        #[test]
        fn test_export_forward_valid() {
            assert_eq!(
                forward("80"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: None,
                            port: 80,
                        }),
                        remote: None,
                    })
                ))
            );
            assert_eq!(
                forward("1.2.3.4:80"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: Some(Host::IP4([1, 2, 3, 4])),
                            port: 80,
                        }),
                        remote: None,
                    })
                ))
            );
            assert_eq!(
                forward("80~80"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: None,
                            port: 80,
                        }),
                        remote: Some(ExportRemoteAddr {
                            onion_alias: None,
                            ports: vec![80],
                        }),
                    })
                ))
            );
            assert_eq!(
                forward("80~80,81,8080,28000"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: None,
                            port: 80,
                        }),
                        remote: Some(ExportRemoteAddr {
                            onion_alias: None,
                            ports: vec![80, 81, 8080, 28000],
                        }),
                    })
                ))
            );
            assert_eq!(
                forward("80~mastodon:80,81,8080,28000"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: None,
                            port: 80,
                        }),
                        remote: Some(ExportRemoteAddr {
                            onion_alias: Some("mastodon".to_string()),
                            ports: vec![80, 81, 8080, 28000],
                        }),
                    })
                ))
            );
            assert_eq!(
                forward("0.0.0.0:80~mastodon:80,81,8080,28000"),
                Ok((
                    "",
                    Forward::Export(ExportForward {
                        local: ExportLocalAddr::TCP(ExportLocalTCPAddr {
                            host: Some(Host::IP4([0, 0, 0, 0])),
                            port: 80,
                        }),
                        remote: Some(ExportRemoteAddr {
                            onion_alias: Some("mastodon".to_string()),
                            ports: vec![80, 81, 8080, 28000],
                        }),
                    })
                ))
            );
        }

        #[test]
        fn test_export_forward_invalid() {
            assert_eq!(
                forward("1.2.3.4"),
                Err(NomErr::Error(VerboseError {
                    errors: vec![
                        (".2.3.4", VerboseErrorKind::Nom(ErrorKind::OneOf)),
                        (".2.3.4", VerboseErrorKind::Nom(ErrorKind::ManyMN)),
                        ("1.2.3.4", VerboseErrorKind::Context("port")),
                        ("1.2.3.4", VerboseErrorKind::Context("local tcp addr")),
                        ("1.2.3.4", VerboseErrorKind::Context("export forward")),
                        ("1.2.3.4", VerboseErrorKind::Nom(ErrorKind::Alt)),
                        ("1.2.3.4", VerboseErrorKind::Context("forward"))
                    ]
                }))
            );
            // TODO: improve nom error messages, they're kinda terrible
            assert_eq!(
                "80,81,82".parse::<Forward>(),
                Err(ParseError(
                    "Parse error:\nEof at: ,81,82\nin section 'forward', at: 80,81,82\n"
                        .to_string()
                ))
            );
            assert_eq!(
                "80,81,82~8000,8001,8002".parse::<Forward>(),
                Err(ParseError("Parse error:\nEof at: ,81,82~8000,8001,8002\nin section 'forward', at: 80,81,82~8000,8001,8002\n".to_string()))
            );
            assert_eq!(
                "".parse::<Forward>(),
                Err(ParseError(
                        "Parse error:\nOneOf at: \nManyMN at: \nin section 'port', at: \nin section 'local tcp addr', at: \nin section 'export forward', at: \nAlt at: \nin section 'forward', at: \n".to_string()
                ))
            );
            assert_eq!(
                "10.0.0.1:8080~192.168.1.1:8080".parse::<Forward>(),
                Err(ParseError("Parse error:\nEof at: .168.1.1:8080\nin section 'forward', at: 10.0.0.1:8080~192.168.1.1:8080\n".to_string()))
            );
        }

        #[test]
        fn test_import_forward_valid() {
            assert_eq!(
                "xyz123.onion".parse::<Forward>(),
                Ok(Forward::Import(ImportForward {
                    remote: ImportRemoteAddr {
                        onion: "xyz123".to_string(),
                        port: None,
                    },
                    local: None,
                }))
            );
            assert_eq!(
                "xyz123.onion:9001".parse::<Forward>(),
                Ok(Forward::Import(ImportForward {
                    remote: ImportRemoteAddr {
                        onion: "xyz123".to_string(),
                        port: Some(9001),
                    },
                    local: None,
                }))
            );
            assert_eq!(
                "xyz123.onion:9001~9002".parse::<Forward>(),
                Ok(Forward::Import(ImportForward {
                    remote: ImportRemoteAddr {
                        onion: "xyz123".to_string(),
                        port: Some(9001),
                    },
                    local: Some(ImportLocalAddr::TCP(ImportLocalTCPAddr {
                        host: None,
                        port: Some(9002),
                    })),
                }))
            );
            assert_eq!(
                "xyz123.onion:9001~172.18.0.1".parse::<Forward>(),
                Ok(Forward::Import(ImportForward {
                    remote: ImportRemoteAddr {
                        onion: "xyz123".to_string(),
                        port: Some(9001),
                    },
                    local: Some(ImportLocalAddr::TCP(ImportLocalTCPAddr {
                        host: Some(Host::IP4([172, 18, 0, 1])),
                        port: None,
                    })),
                }))
            );
        }

        #[test]
        fn test_import_forward_invalid() {
            assert_eq!(
                "xyz123".parse::<Forward>(),
                Err(ParseError("Parse error:\nOneOf at: xyz123\nManyMN at: xyz123\nin section 'port', at: xyz123\nin section 'local tcp addr', at: xyz123\nin section 'export forward', at: xyz123\nAlt at: xyz123\nin section 'forward', at: xyz123\n".to_string()))
            );
            assert_eq!(
                "xyz123.shallot".parse::<Forward>(),
                Err(ParseError("Parse error:\nOneOf at: xyz123.shallot\nManyMN at: xyz123.shallot\nin section 'port', at: xyz123.shallot\nin section 'local tcp addr', at: xyz123.shallot\nin section 'export forward', at: xyz123.shallot\nAlt at: xyz123.shallot\nin section 'forward', at: xyz123.shallot\n".to_string()))
            );
            assert_eq!(
                "xyz123.onion~abc123.onion".parse::<Forward>(),
                Err(ParseError("Parse error:\nEof at: ~abc123.onion\nin section 'forward', at: xyz123.onion~abc123.onion\n".to_string()))
            );
        }
    }
}
